python/uv/private/lock.bzl

```diff
@@ -15,109 +15,242 @@
+    user_args = args
+    args = [
         "--generate-hashes",
         "--emit-index-url",
         "--no-strip-extras",
-        "--python=$(PYTHON3)",
-    ] + args + [
-        "$(location {})".format(src)
-        for src in srcs
     ]
-    if upgrade:
-        _args.append("--upgrade")
-    if universal:
-        _args.append("--universal")
-    _args.append("--output-file=$@")
-    cmd = "$(UV_BIN) pip compile " + " ".join(_args)
-
-    # Make a copy to ensure that we are not modifying the initial list
-    srcs = list(srcs)
-
-    # Check if the output file already exists, if yes, first copy it to the
-    # output file location in order to make `uv` not change the requirements if
-    # we are just running the command.
-    if native.glob([out]):
-        cmd = "cp -v $(location {}) $@; {}".format(out, cmd)
-        srcs.append(out)
-
-    native.genrule(
+    args += user_args
+    maybe_out = _maybe_path(out)
+
+    _lock(
         name = name,
         srcs = srcs,
-        outs = [out + ".new"],
-        cmd_bash = cmd,
+        # Check if the output file already exists, if yes, first copy it to the
+        # output file location in order to make `uv` not change the requirements if
+        # we are just running the command.
+        maybe_out = maybe_out,
+        out = out + ".new",
         tags = [
             "local",
             "manual",
             "no-cache",
         ],
+        args = args,
         target_compatible_with = _REQUIREMENTS_TARGET_COMPATIBLE_WITH,
-        toolchains = [
-            Label("//python/uv:current_toolchain"),
-            Label("//python:current_py_toolchain"),
-        ],
     )
 
+    _run_lock(
+        name = locker_target,
+        lock = name,
+    )
+
+    if maybe_out:
+        diff_test(
+            name = name + "_test",
+            file1 = out + ".new",
+            file2 = maybe_out,
+            tags = ["manual"],
+        )
+
     # Write a script that can be used for updating the in-tree version of the
     # requirements file
-    write_file(
-        name = name + ".update_gen",
+    expand_template(
+        name = update_target + "_gen",
         out = update_target + ".py",
-        content = [
-            "from os import environ",
-            "from pathlib import Path",
-            "from sys import stderr",
-            "",
-            'src = Path(environ["REQUIREMENTS_FILE"])',
-            'assert src.exists(), f"the {src} file does not exist"',
-            'dst = Path(environ["BUILD_WORKSPACE_DIRECTORY"]) / "{}" / "{}"'.format(pkg, out),
-            'print(f"Writing requirements contents\\n  from {src.absolute()}\\n  to {dst.absolute()}", file=stderr)',
-            "dst.write_text(src.read_text())",
-            'print("Success!", file=stderr)',
-        ],
+        template = "//python/uv/private:copy.py",
+        substitutions = {
+            'dst = ""': 'dst = "{}/{}"'.format(pkg, out),
```

Comment:
This is emulating `$(rootpath out)`.

I think we could extract a macro for writing files into the source tree.